README
README file for CGI::Snapp::Plugin::Redirect.

See also: CHANGES and Changelog.ini.

Warning: WinZip 8.1 and 9.0 both contain an 'accidental' bug which stops
them recognizing POSIX-style directory structures in valid tar files.
You are better off using a reliable tool such as InfoZip:
ftp://ftp.info-zip.org/pub/infozip/

1 Installing from a Unix-like distro
------------------------------------
shell>gunzip CGI-Snapp-Plugin-Redirect-1.00.tgz
shell>tar mxvf CGI-Snapp-Plugin-Redirect-1.00.tar

On Unix-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>./Build
shell>./Build test
shell>./Build install

On MS Windows-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>perl Build
shell>perl Build test
shell>perl Build install

Alternately, without Module::Build, you do this:

Note: 'make' on MS Windows-like systems may be called 'nmake' or 'dmake'.

shell>perl Makefile.PL
shell>make
shell>make test
shell>su              (for Unix-like systems)
shell>make install
shell>exit            (for Unix-like systems)

On all systems:

Run Redirect.pm through you favourite pod2html translator.

2 Installing from an ActiveState distro
---------------------------------------
shell>unzip CGI-Snapp-Plugin-Redirect-1.00.zip
shell>ppm install --location=. CGI-Snapp-Plugin-Redirect
shell>del CGI-Snapp-Plugin-Redirect-1.00.ppd
shell>del PPM-CGI-Snapp-Plugin-Redirect-1.00.tar.gz


CHANGES
Revision history for Perl extension CGI::Snapp::Plugin::Redirect.

1.01  Mon Apr  9 17:00:00 2012
	- Change tests to match CGI::Snapp V 1.01 which does not create a logger by default.

1.00  Mon Feb 28 17:00:00 2012
	  - Original version


t/lib/CGI/Snapp/Plugin/RedirectTest.pm
package CGI::Snapp::Plugin::RedirectTest;

use parent 'CGI::Snapp';
use strict;
use warnings;

use CGI::Snapp::Plugin::Redirect;

our $VERSION = '1.01';

# --------------------------------------------------

sub cgiapp_prerun
{
	my($self) = @_;
	my($name) = __PACKAGE__ . '.cgiapp_prerun()';

	print "$name\n";

	$self -> redirect('http://first.net.au/') if ($self -> param('test.prerun.mode') );

} # End of cgiapp_prerun.

# --------------------------------------------------

sub first_sub
{
	my($self) = @_;
	my($name) = __PACKAGE__ . '.first_sub()';

	print "$name\n";

	if ($self -> param('test.local.url') )
	{
		$self -> redirect('login.html');
	}
	elsif ($self -> param('test.without.status') )
	{
		$self -> redirect('http://second.net.au/');
	}
	elsif ($self -> param('test.with.status') )
	{
		$self -> redirect('http://third.net.au/', '301 Moved Permanently');
	}

	return $name;

} # End of first_sub.

# --------------------------------------------------

sub setup
{
	my($self) = @_;
	my($name) = __PACKAGE__ . '.setup()';

	print "$name\n";

	# Add callback methods belonging to this package.

	__PACKAGE__ -> add_callback('prerun', 'cgiapp_prerun');

} # End of setup.

# --------------------------------------------------

1;


Makefile.PL
use ExtUtils::MakeMaker;

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

WriteMakefile
(
 ($] ge '5.005') ?
 (
  AUTHOR   => 'Ron Savage (ron@savage.net.au)',
  ABSTRACT => 'A plugin for CGI::Snapp to simplify using HTTP redirects',
    ) : (),
 clean =>
 {
	 FILES => 'blib/* Makefile MANIFEST CGI-Snapp-Plugin-Redirect-*'
 },
 dist =>
 {
	 COMPRESS => 'gzip',
	 SUFFIX   => 'gz'
 },
 DISTNAME  => 'CGI-Snapp-Plugin-Redirect',
 NAME      => 'CGI::Snapp::Plugin::Redirect',
 PL_FILES  => {},
 PREREQ_PM =>
 {
	CGI          => 3.59,
	CGI::Snapp   => 1.01,
	Exporter     => 5.567,
	Log::Handler => 0.75,
	parent       => 0,
	strict       => 1.03,
 	Test::More   => 0.94,
 	Test::Pod    => 1.44,
	vars         => 1.01,
	warnings     => 1.03,
 },
 VERSION_FROM => 'lib/CGI/Snapp/Plugin/Redirect.pm',
);


lib/CGI/Snapp/Plugin/Redirect.pm
package CGI::Snapp::Plugin::Redirect;

use strict;
use warnings;

use vars qw(@EXPORT @ISA);

@EXPORT = ('redirect');
@ISA    = ('Exporter');

our $VERSION = '1.01';

# --------------------------------------------------

sub redirect
{
	my($self, $url, $status) = @_;
	$url    ||= '';
	$status ||= 0;

	$self -> log(debug => "redirect($url, ...)");

	# If we're in the prerun phase, generate a no-op via a dummy sub.

	if ($self -> _prerun_mode_lock == 0)
	{
		$self -> run_modes(dummy_redirect => sub{});
		$self -> prerun_mode('dummy_redirect');
	}

	if ($status)
	{
		$self -> header_add(-location => $url, -status => $status);
	}
	else
	{
		$self -> header_add(-location => $url);
	}

	$self -> header_type('redirect');

} # End of redirect.

# --------------------------------------------------

1;

=pod

=head1 NAME

CGI::Snapp::Plugin::Redirect - A plugin for CGI::Snapp to simplify using HTTP redirects

=head1 Synopsis

	package My::App;

	use parent 'CGI::Snapp';

	use CGI::Snapp::Plugin::Redirect;

	sub cgiapp_prerun
	{
		my($self) = @_;

		if (<< not logged in >>)
		{
			$self -> redirect('login.html');
		}

	} # End of cgiapp_prerun.

	sub teardown
	{
		my($self) = @_;

		return $self -> redirect('http://www.example.com/', '301 Moved Permanently');

	} # End of teardown.

=head1 Description

When you 'use' this module in your sub-class of L<CGI::Snapp> (as in the Synopsis), it automatically imports into your sub-class the L</redirect($url[, $status])> method, to give you
a single call to set the HTTP headers for redirection to an external $url. See that method's details below for exactly what effect a call to redirect() has.

If you just want to display the results of another run mode within the same application, then L<CGI::Snapp::Plugin::Forward>'s forward() method is more suitable.

=head1 Distributions

This module is available as a Unix-style distro (*.tgz).

See L<http://savage.net.au/Perl-modules/html/installing-a-module.html>
for help on unpacking and installing distros.

=head1 Installation

Install L<CGI::Snapp::Plugin::Redirect> as you would for any C<Perl> module:

Run:

	cpanm CGI::Snapp::Plugin::Redirect

or run:

	sudo cpan CGI::Snapp::Plugin::Redirect

or unpack the distro, and then either:

	perl Build.PL
	./Build
	./Build test
	sudo ./Build install

or:

	perl Makefile.PL
	make (or dmake or nmake)
	make test
	make install

=head1 Constructor and Initialization

This module does not have, and does not need, a constructor.

=head1 Methods

=head2 redirect($url[, $status])

Interrupts the current request, and redirects to the given (external) $url, optionally setting the HTTP status to $status.

Here, the [] indicate an optional parameter.

The redirect happens even if you are inside a method attached to the 'prerun' hook when you call redirect().

Specifically, this method does these 3 things:

=over 4

=item o Sets the HTTP header 'location' to the given $url

=item o Sets the HTTP 'status' (if provided) to $status

=item o Sets the L<CGI::Snapp> header type to 'redirect'

=back

=head1 FAQ

=head2 Why don't you 'use Exporter;'?

It is not needed; it would be for documentation only.

For the record, Exporter V 5.567 ships with Perl 5.8.0. That's what I had in Build.PL and Makefile.PL until I tested the fact I can omit it.

=head1 See Also

L<CGI::Application>

The following are all part of this set of distros:

L<CGI::Snapp> - A almost back-compat fork of CGI::Application

L<CGI::Snapp::Dispatch> - Dispatch requests to CGI::Snapp-based objects

L<CGI::Snapp::Plugin::Forward> - A plugin for CGI::Snapp to switch cleanly to another run mode within the same app

L<CGI::Snapp::Plugin::Redirect> - A plugin for CGI::Snapp to simplify using HTTP redirects

L<CGI::Snapp::Demo::One> - A template-free demo of CGI::Snapp using just 1 run mode

L<CGI::Snapp::Demo::Two> - A template-free demo of CGI::Snapp using N run modes

L<CGI::Snapp::Demo::Three> - A template-free demo of CGI::Snapp using CGI::Snapp::Plugin::Forward

L<CGI::Snapp::Demo::Four> - A template-free demo of CGI::Snapp using Log::Handler::Plugin::DBI

L<CGI::Snapp::Demo::Four::Wrapper> - A wrapper around CGI::Snapp::Demo::Four, to simplify using Log::Handler::Plugin::DBI

L<Config::Plugin::Tiny> - A plugin which uses Config::Tiny

L<Config::Plugin::TinyManifold> - A plugin which uses Config::Tiny with 1 of N sections

L<Data::Session> - Persistent session data management

L<Log::Handler::Plugin::DBI> - A plugin for Log::Handler using Log::Hander::Output::DBI

L<Log::Handler::Plugin::DBI::CreateTable> - A helper for Log::Hander::Output::DBI to create your 'log' table

=head1 Machine-Readable Change Log

The file CHANGES was converted into Changelog.ini by L<Module::Metadata::Changes>.

=head1 Version Numbers

Version numbers < 1.00 represent development versions. From 1.00 up, they are production versions.

=head1 Credits

Please read L<https://metacpan.org/module/CGI::Application::Plugin::Redirect#AUTHOR>, since this code is basically copied from L<CGI::Application::Plugin::Redirect>.

=head1 Support

Email the author, or log a bug on RT:

L<https://rt.cpan.org/Public/Dist/Display.html?Name=CGI::Snapp::Plugin::Redirect>.

=head1 Author

L<CGI::Snapp::Plugin::Redirect> was written by Ron Savage I<E<lt>ron@savage.net.auE<gt>> in 2012.

Home page: L<http://savage.net.au/index.html>.

=head1 Copyright

Australian copyright (c) 2012, Ron Savage.

	All Programs of mine are 'OSI Certified Open Source Software';
	you can redistribute them and/or modify them under the terms of
	The Artistic License, a copy of which is available at:
	http://www.opensource.org/licenses/index.html

=cut
